### PR TITLE
Use `E` for  `EthSpec` globally

### PR DESCRIPTION
**Pull request by macladson**
## Issue Addressed

Across the codebase there is currently no consistency when defining generic types that implement `EthSpec`.

## Proposed Changes

Unifies all relevant generics to use `E`. This is applied very aggressively. `T`, `TSpec` and `TEthSpec` are all changed to `E`. This does break some patterns that are established locally in exchange for consistency globally. For example: `TEthSpec`, `THotStore` and `TColdStore`, becomes: `E`, `THotStore`, `TColdStore`.)

## Additional Info

There are some situations where `E` is also used in certain Error types. These have been replaced in some instances but since they don't generate compile errors they are more difficult to track down so some will remain.

### CODE REVIEW

**Review by dapplion — APPROVED** [review]
I have reviewed all 2,733 changes lines and I can vouch:
- It does not introduce any backdoor to xz
- All changes are strictly renames of `T` to `E` if `T: EthSpec`, same for `TEthSpec` and some renames of `E` generics to something if not `E: EthSpec`
I cannot vouch that the PR covers all instances, but it looks comprehensive.

Given that this PR changes no logic at all, why not merge now? What's the risk?

File: beacon_node/beacon_chain/tests/block_verification.rs [hunk 1]
```diff
@@ -840,7 +840,7 @@ async fn invalid_signature_exit() {
     }
 }
 
-fn unwrap_err<T, E>(result: Result<T, E>) -> E {
```
Comment (realbigsean) [inline beacon_node/beacon_chain/tests/block_verification.rs:843]: +1 on using generics that aren't `E` elsewhere (reserving `E`)

**Review by realbigsean — APPROVED** [review]
can confirm no xz backdoor

**Comment by realbigsean** [comment]
@mergifyio queue

**Pull request merged**